### PR TITLE
Allow assert.Equal on string type alias without panicking on failure

### LINKED ISSUE #644: Panic: Equal on type aliases causes panic
Current behaviour:

With go 1.9.3 and testify 1.2.2 or latest master (f35b8ab0b5a2cef36673838d662e249dd9c94686)  if I require/assert.Equal two aliased types, it causes testify to panic:

```
goroutine 5 [running]:
testing.tRunner.func1(0xc42010a0f0)
	/Users/brendan.fosberry/.gvm/gos/go1.9.3/src/testing/testing.go:711 +0x2d2
panic(0x1289000, 0xc420066780)
	/Users/brendan.fosberry/.gvm/gos/go1.9.3/src/runtime/panic.go:491 +0x283
github.com/stretchr/testify/assert.diff(0x126fae0, 0xc420010d10, 0x126fae0, 0x130a0b0, 0x0, 0x0)
	/Users/brendan.fosberry/go/src/github.com/stretchr/testify/assert/assertions.go:1352 +0x494
github.com/stretchr/testify/assert.Equal(0x141c480, 0xc42010a0f0, 0x126fae0, 0xc420010d10, 0x126fae0, 0x130a0b0, 0x0, 0x0, 0x0, 0xc420039f01)
	/Users/brendan.fosberry/go/src/github.com/stretchr/testify/assert/assertions.go:338 +0x271
_/Users/brendan.fosberry/workspace/derp.TestMyTypeRequire(0xc42010a0f0)
	/Users/brendan.fosberry/workspace/derp/derp_test.go:19 +0x11f
testing.tRunner(0xc42010a0f0, 0x12e9c88)
	/Users/brendan.fosberry/.gvm/gos/go1.9.3/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/Users/brendan.fosberry/.gvm/gos/go1.9.3/src/testing/testing.go:789 +0x2de
FAIL	_/Users/brendan.fosberry/workspace/derp	0.038s
```

Expected behaviour: testify should determine if the two types are equal, or fail gracefully.

Here is a test case to replicate the issue

```
package main

import (
        "testing"

        "github.com/stretchr/testify/assert"
)

type MyType string

const (
        MyTypeA MyType = "a"
        MyTypeB MyType = "b"
)

func TestMyTypeRequire(t *testing.T) {
        mt := MyTypeA
        assert.Equal(t, mt, MyTypeA)
        assert.Equal(t, mt, MyTypeB)
}
```

I don't know much about reflect, but I think it comes down to this:

```
	var e, a string
	if ek != reflect.String {
		e = spewConfig.Sdump(expected)
		a = spewConfig.Sdump(actual)
	} else {
		e = expected.(string)
		a = actual.(string)
	}
```

ek == String comes back as true for string type aliases, however a type assertion causes a panic since they are of type string.

My workaround for now is not to use Equal with type aliases, however this is something that should be handled internally.

### LINKED ISSUE #646: require.Equal panics when using string type alias
I'm trying to compare IDs of package github.com/graph-gophers/graphql-go. These IDs are actually just a type alias for strings:
```go
// ID represents GraphQL's "ID" scalar type. A custom type may be used instead.
type ID string
```

However if these IDs are not equal, require.Equal panics:
```go
func TestResolver(t *testing.T) {
	require.Equal(t, graphql.ID("2015"), graphql.ID("1915"))
}
```
```bash
=== RUN   TestResolver
--- FAIL: TestResolver (0.00s)
panic: interface conversion: interface {} is graphql.ID, not string [recovered]
	panic: interface conversion: interface {} is graphql.ID, not string

goroutine 148 [running]:
testing.tRunner.func1(0xc4203ac1e0)
	/usr/local/go/src/testing/testing.go:742 +0x567
panic(0x1a60400, 0xc420280440)
	/usr/local/go/src/runtime/panic.go:502 +0x24a
github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/assert.diff(0x1a56c80, 0x1c484e0, 0x1a56c80, 0x1c484f0, 0x0, 0x0)
	/Users/matthias/Golang/src/github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/assert/assertions.go:1352 +0x4ba
github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/assert.Equal(0x44a0188, 0xc4203ac1e0, 0x1a56c80, 0x1c484e0, 0x1a56c80, 0x1c484f0, 0x0, 0x0, 0x0, 0xc4203ac1e0)
	/Users/matthias/Golang/src/github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/assert/assertions.go:338 +0x2f8
github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/require.Equal(0x1c508e0, 0xc4203ac1e0, 0x1a56c80, 0x1c484e0, 0x1a56c80, 0x1c484f0, 0x0, 0x0, 0x0)
	/Users/matthias/Golang/src/github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/require/require.go:157 +0xc0
github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/require.(*Assertions).Equal(0xc42036ae68, 0x1a56c80, 0x1c484e0, 0x1a56c80, 0x1c484f0, 0x0, 0x0, 0x0)
	/Users/matthias/Golang/src/github.com/SpaceHeroGuide/wertungsheft/vendor/github.com/stretchr/testify/require/require_forward.go:130 +0xe3
github.com/SpaceHeroGuide/wertungsheft/pkg/api/tournament_test.TestResolver(0xc4203ac1e0)
	/Users/matthias/Golang/src/github.com/SpaceHeroGuide/wertungsheft/pkg/api/tournament/tournament_test.go:90 +0x163
testing.tRunner(0xc4203ac1e0, 0x1b6a7a0)
	/usr/local/go/src/testing/testing.go:777 +0x16e
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:824 +0x565
FAIL	github.com/SpaceHeroGuide/wertungsheft/pkg/api/tournament	2.198s
```

If you change the IDs so they are equal, the panic does not occur.

### PR DESCRIPTION
**Pull request by HaraldNordgren**
@ernesto-jimenez The old comparison against `reflect.String` is true for type aliases of string, but the type coercion panics. So let's compare the type instead of the kind, to avoid the panics.

Fixes https://github.com/stretchr/testify/issues/644
Fixes https://github.com/stretchr/testify/issues/646

### CODE REVIEW

File: assert/assertions_test.go
Comment (HaraldNordgren) [inline assert/assertions_test.go]: This test will panic if the fix in `assert/assertions.go` is reverted.

**Review by georgelesica-wf — APPROVED** [review]
Looks good, thank you for your patience!

**Comment by HaraldNordgren** [comment]
@ernesto-jimenez @georgelesica-wf Merge this? 😊

**Comment by georgelesica-wf** [comment]
@HaraldNordgren I'm waiting on merge permissions from @matryer - hang tight, sorry!

**Pull request merged**
